clc/modules/object-storage/src/main/java/com/eucalyptus/objectstorage/ObjectStorageGateway.java

```diff
@@ -2013,113 +2055,317 @@ public DeleteBucketTaggingResponseType deleteBucketTagging(DeleteBucketTaggingTy
+        CorsPreflightInvalidMethodException s3e = new CorsPreflightInvalidMethodException(requestMethod);
+        throw s3e;
+      }
+
+      List<CorsRule> corsRules = BucketCorsManagers.getInstance().getCorsRules(bucket.getBucketUuid());
+
+      if (corsRules == null || corsRules.isEmpty()) {
+        CorsPreflightNoConfigException s3e = new CorsPreflightNoConfigException(requestMethod,
+            key == null ? "BUCKET" : "OBJECT");
+        throw s3e;     
+      }
+
+      List<String> requestHeaders = preflightRequest.getRequestHeaders();
+      CorsMatchResult corsMatchResult = OSGUtil.matchCorsRules (corsRules, requestOrigin, requestMethod, requestHeaders);
+      
+      CorsRule corsRuleMatch = corsMatchResult.getCorsRuleMatch();
+      if (corsRuleMatch == null) {
+        // No rule matched the request
+        CorsPreflightNotAllowedException s3e = new CorsPreflightNotAllowedException(requestMethod,
+            key == null ? "BUCKET" : "OBJECT");
+        throw s3e;     
+      }
+      
+      // We found a match, fill in the response fields
+      PreflightResponse responseFields = new PreflightResponse();
+      response.setPreflightResponse(responseFields);
+      
+      // If the origin we matched against is "*" (any origin allowed), then
+      // set the response's origin to "*" instead of the origin in the 
+      // request. Matches AWS behavior and W3 spec: (strange but true)
+      // https://www.w3.org/TR/cors/#resource-preflight-requests
+      responseFields.setOrigin(corsMatchResult.getAnyOrigin() ? "*" : requestOrigin);
+
+      // Return all the CORS rule's allowed methods in the response
+      responseFields.setMethods(corsRuleMatch.getAllowedMethods());
+      
+      // Return all of the request's Access-Control-Request-Headers ni the response
+      responseFields.setAllowedHeaders(preflightRequest.getRequestHeaders());
+
+      // Return all of the CORS rule's expose headers in the response
+      responseFields.setExposeHeaders(corsRuleMatch.getExposeHeaders());
+      
+      responseFields.setMaxAgeSeconds(corsRuleMatch.getMaxAgeSeconds());
+
+      response.setStatus(HttpResponseStatus.OK);
+
+    } catch (S3Exception s3e) {
+      LOG.warn("Caught S3Exception while processing the preflight CORS request for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with: ", s3e);
+      throw s3e;
+    } catch (Exception ex) {
+      LOG.warn("Caught general exception while processing the preflight CORS request for bucket <" + 
+          bucketName + ">, CorrelationId: " + Contexts.lookup().getCorrelationId() + 
+          ", responding to client with 500 InternalError because of: ", ex);
+      throw new InternalErrorException(bucketName, ex);
+    }
+    
+    return response;
+  }
+
+  public static void addCorsResponseHeaders (MappingHttpResponse httpResponse) throws S3Exception {
```

Comment:
Refactor static methods employed by pipeline stages and handlers to a utility class such as OSGUtil.java. ObjectStorageGateway.java should be treated as the implementation of the service API - ObjectStorageService. Its okay to have private static methods for use within this class. Defining public static methods renders the ability to invoke this service implementation without the necessity to instantiate (configure()) and exposes unintended functionality. In addition the flow is counter intuitive when a pipeline stage such as the outbound handler invokes functionality encapsulated a previous pipeline stage